File: python/src/strings/valid_anagram.py
"""True if two strings are anagrams (same multiset of characters)."""


def valid_anagram(a: str, b: str) -> bool:
    if len(a) != len(b):
        return False
    counts: dict[str, int] = {}
    for c in a:
        counts[c] = counts.get(c, 0) + 1
    for c in b:
        if c not in counts:
            return False
        counts[c] -= 1
        if counts[c] == 0:
            del counts[c]
    return len(counts) == 0

File: typescript/src/strings/validAnagram.ts
export function validAnagram(a: string, b: string): boolean {
  if (a.length !== b.length) return false;
  const counts = new Map<string, number>();
  for (let i = 0; i < a.length; i++) {
    const ca = a[i]!;
    const cb = b[i]!;
    counts.set(ca, (counts.get(ca) ?? 0) + 1);
    counts.set(cb, (counts.get(cb) ?? 0) - 1);
  }
  for (const v of counts.values()) {
    if (v !== 0) return false;
  }
  return true;
}
